src/port_scan/tcp_scan.1
.TH TCP_SCAN 1
.SH NAME
tcp_scan, udp_scan \- internet port scanners
.SH SYNOPSIS
.B tcp_scan
[-abuU] [-l load] [-s string] [-w time] host service(s)...
.sp
.B udp_scan
[-apuU] [-l load] host service(s)...
.SH DESCRIPTION
These commands take a list of internet services and investigate which
services are available from a given host.  The \fItcp_scan\fR command
looks for connection-oriented services; \fIudp_scan\fR identifies
active datagram ports.
.sp
Each \fIservice\fR argument may be specified as a symbolic name
(telnet), a port number (23), an interval (1-1023, telnet-smtp) or an
interval with the lower or upper bounds missing (the default bounds are
1 and 65535, respectively).
.sp
Options:
.IP -a
Report status of all specified services.
.IP "-b (tcp_scan only)"
Report banner information. Banners are converted to printable form
using C-like escape sequences. Whenever \fItcp_scan\fR finds that the
server does telnet options negotiation it sets the 't' flag in the
output.  This is useful to detect telnet servers on non-standard ports.
.IP "-l load"
Minimize the impact of network roundtrip delays by performing
\fIload\fR network probes in parallel. The default load is the
per-process open-file limit - 10.
.IP "-p port (udp_scan only)"
Use this port to verify that the host or network is alive. By default
the UDP port scanner uses port number 1. Specify a port number that is
known to be unreachable or inactive.
.IP "-s string (tcp_scan only)"
After a connection has been established, send
.I string
to the server. The following backslash escapes can be used:  \\ooo
(octal character code), \\b (backspace), \\f (formfeed), \\n (newline),
\\r (carriage-return), \\s (space), and \\t (horizontal tab). The
.I string
should be enclosed between quotes if it contains shell meta characters.
.IP "-t time (tcp_scan only)"
Give up when the program has not found out anything within
.I time
seconds. This bounds the time lost when scanning systems with broken
TCP/IP implementations that do not send RESETs when contacted at a dead
port.  The udp scanner already has a built-in mechanism to detect dead
hosts.
.IP -u
Report probes that fail with "Host unreachable". Use this with packet
filters that pass most traffic.
.IP -U
Report probes that do \fInot\fR fail with "Host unreachable". Use this
with packet filters that block most traffic.
.IP "-w time (tcp_scan only)"
Wait for at most \fItime\fR seconds for banner information.
.SH WARNING
These programs will raise lots of alarms on sites that run the \fItcp
wrapper\fR or other network logging software. Use only with prior
permission.
.SH BUGS
The UDP port scanner relies on ICMP replies to detect that a service is
unavailable, and may report false positives when the host or network
dies in the middle of a measurement.
.PP
The TCP port scanner does not keep track of roundtrip times or
of retransmissions, and may overload hosts or networks.
.PP
With some UNIX implementations, a single "Host unreachable" condition
affects all connections that are being established with that host.
.PP
With some UNIX implementations, a single "Host unreachable" condition
affects all TCP connections to that host, even those that already
exist.
.SH AUTHOR
Wietse Venema


Changes
1.1.1

- URL filter was too strict, would reject names with () in them.

1.1

- bundled in ctime.pl and getopts.pl from the perllib.
- trouble shooting docs now integrated with FAQ, and a full week of feedback
  processed into FAQ and troubkle entries.
- added a 13th vulnerability page. Irony has it that SATAN is the subject.
  Surprise: CERT/CC posts advisory *before* a software update is available.
- SATAN now detects that HTML clients reveal parent URL information.
- eliminated extraneous ping probes.
- strerror() is provided whether or not the compiler understands ANSI.
- prevent compiler warnings about "& before array" in rpcgen output.
- use waitpid() to find out when the HTML browser terminates.
- moved 'require getopts.pl' down in bin/*.satan scripts.
- fixed typos/errors/omissions/ambiguities in the documentation.
- the -c option now changes individual perl variables. Example:
  -c "dont_use_nslookup = 1; dont_use_icmp = 1". Useful for config
  variables that have no command-line option.
- dropped the -c alternate config file feature. There is no easy way
  to both set *and* override defaults (short of writing our own getopt).
- getfqdn() could be blind for nameless hosts when nslookup was enabled.
- getfqdn() would be blind for nameless hosts when nslookup was disabled.
- rsh test no longer relies on the rsh or remsh command (it worked 
  only when the target account exists on the probing host), and it
  now makes difference between "host trusts everyone" and "account
  trusts everyone". It would claim that hosts are open when they
  refuse access with a tcp wrapper banner message.
- append dot in dns.satan query to avoid spurious name servers being
  listed. This also may make things faster.
- avoid duplicate "exports to the world" message (shut up showmount
  message when running from an untrusted host).
- added HTML client address check, in case the magic cookie leaks out.
  This is just a last barrier; cookies should never be disclosed.
- added URL name address check, in case the magic cookie leaks out.
  This is just a last barrier; cookies should never be disclosed.
- added an ANSI/POSIXized version of the rpcgen command for systems
  without one (ultrix, some sysv4 versions).
- more Linux support. There are many Linux versions so who knows where
  this will work.
- rsh tests are now executed only when the probing host is untrusted.
- avoid duplicate tftp problem message by making it different for read 
  and write.

1.0! We made it!

- just in case, dropped TUE references from all code except safe_finger
  (which was stolen from the TCP wrapper) and the admin guide (which
  was already posted one year ago with TUE reference in it).
- dropped the unused rcmdinfo tool.
- linked the troubleshooting guide.
- deleted all references to set-uid tools - scripts cannot be made setuid.
- added note about parallel runs to the satan.8 man page.
- will ask the user to set $dont_use_dns when nslookup fails.
- README & TODO - added disclaimer on parallelism; added AFS & SNMP to our
  TODO list.
- rm'd fireup
- added proxy stuff to README and troubleshooting.html

0.93

- minor typos in the documentation keep popping up...
- worked around a sys_errlist[] declaration clash in freebsd 2.0.
- changed references to "session key" to "magic cookie".
- fixed "back to.." link in the system requirements.
- gave the session key writeup a more architectural tone.
- fixed another duplicate filtering bug in danger report.
- SATAN no longer dies (undefined host) when given an incomplete database.
- clearer distinction of in-core and on-disk databases, and what happens
  when you load/merge/open/query them.
- fixed some typos in the FAQ.
- commented out offending commands in satan.ps
- fixed some typos in sendmail check.

0.91

- docs: user_interface, the_main_parts, satan_reference, satan_overview,
  satan.db, FAQ (incl. adding your stuff to FAQ)
- rules - 2 more sendmail checks - (<= 5.60), (<=5.65 && dynix) in facts,
  improved DYNIX recognition
- changed tutorial Sendmail_vulnerabilities.html

0.90

- more consistency (use of SATAN logo, intermediate headings).
- broke down documentation into overview and reference.
- fixed some dead links (_ instead of -).
- figured out why lynx would not start up a problm.

- put back in old/new satan.ps
- writable-FTP tutorial - added cert link
- reconfig - changed error msg to be more clear
- minor changes to README
- minor grammar and spelling changes to satan.8
- Changed several docs - clarity, expansion, etc.
- put in the troubleshooting.html guide

- figured out why nfs-chk/yp-chk had trouble with Solaris.

- cloned nfs-chk.satan to yp-chk.satan for nis map accessibility test.
- added -d (domain name) option to ypbind.satan.
- cloned nfs-chk to yp-chk and plugged in an YP client call.
- Linux users are now instructed to copy BSD includes into the satan tree.

- found the nfs ghost - missing initialization in dynamic page.
- why tcpscan was too quick claimin telnet on non-std port.
- added <HTML><HEAD><BODY> to the dynamic HTML pages.
- consistent spelling of names (caps).
- copied the alpha satan.ps, the ./satan.ps was busted.
- changed font in acknowledgements from fixed width to something more
  palatable.

- bold faced our names in authors.html
- fixed satan => SATAN in satan.8
- added satan.ps
- modified/added to html/docs/FAQ.html (red 'n' black dot controversy
  on a B/W screen, changing HTML viewers, etc.)
- mv'd html/docs/admin_guide_to_cracking => $!.html
- changed a link in html/docs/references.html to above
- changed html/docs/user_interface.html to have ALT's to dots
- put a <HTML>, <HEAD>, and <BODY> into html/docs/*.html, and
  in html/tutorials/vulnerability/*, also added some "ALT=..."
- put in (identical) warnings about tcp-wrappers/reverse fingers, in
  running for the first time tutorial and general usage docs.
- nuked the glossary - no time to finish it right.
- Nuked *.orig, *.bak, *.old


src/port_scan/tcp_scan.c
 /*
  * tcp_scan - determine available tcp services, optionally collect banners
  * and detect telnet options
  * 
  * Author: Wietse Venema.
  */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/time.h>
#include <sys/socket.h>

#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <netinet/tcp.h>

#include <arpa/telnet.h>

#include <stdio.h>
#include <signal.h>
#include <netdb.h>
#include <string.h>
#include <errno.h>

extern int errno;
extern char *optarg;
extern int optind;

#ifndef __STDC__
extern char *strerror();
#endif

#define offsetof(t,m)	(size_t)(&(((t *)0)->m))

#ifndef IPPORT_TELNET
#define IPPORT_TELNET	23
#endif

#ifndef FD_SET
#include <sys/select.h>
#endif

#include "lib.h"

#define BANNER_LENGTH	2048		/* upper bound on banner info */
#define BANNER_TIME	10		/* time for host to send banner */
#define BANNER_IDLE	1		/* delay after last banner info */

#define YES     1
#define NO      0

#define WAIT	1
#define NOWAIT	0

int     verbose;			/* default silent mode */
int     banner_time = BANNER_TIME;	/* banner timeout */
int     open_file_limit;		/* max nr of open files */
int     load_limit;			/* max nr of open sockets */

struct timeval now;			/* banner_info last update time */
fd_set  write_socket_mask;		/* sockets with connect() in progress */
fd_set  read_socket_mask;		/* sockets with connect() finished */
int     ports_busy;			/* number of open sockets */
int     ports_done;			/* number of finished sockets */
int     max_sock;			/* max socket file descriptor */
int     want_err;			/* want good/bad news */
int     show_all;			/* report all ports */

int    *socket_to_port;			/* socket to port number */

typedef struct {
    unsigned char *buf;			/* banner information or null */
    int     count;			/* amount of banner received sofar */
    int     flags;			/* see below */
    struct timeval connect_time;	/* when connect() finished */
    struct timeval read_time;		/* time of last banner update */
} BANNER_INFO;

BANNER_INFO *banner_info = 0;

#define F_TELNET	(1<<0)		/* telnet options seen */

int     icmp_sock;			/* for unreachable reports */
static struct sockaddr_in sin;		/* remote endpoint info */

static char *send_string;		/* string to send */
int response_time;			/* need some response */

#define NEW(type, count) (type *) mymalloc((count) * sizeof(type))

#define time_since(t) (now.tv_sec - t.tv_sec + 1e-6 * (now.tv_usec - t.tv_usec))

/* main - command-line interface */

main(argc, argv)
int     argc;
char   *argv[];
{
    int     c;
    struct protoent *pe;
    char  **ports;

    progname = argv[0];
    if (geteuid())
	error("This program needs root privileges");

    open_file_limit = open_limit();
    load_limit = open_file_limit - 10;

    while ((c = getopt(argc, argv, "abl:s:t:uUvw:")) != EOF) {
	switch (c) {
	case 'a':
	    show_all = 1;
	    break;
	case 'b':
	    if (banner_info == 0)
		banner_info = NEW(BANNER_INFO, open_file_limit);
	    break;
	case 'l':
	    if ((load_limit = atoi(optarg)) <= 0)
		usage("invalid load limit");
	    if (load_limit > open_file_limit - 10)
		load_limit = open_file_limit - 10;
	    break;
	case 's':
	    send_string = optarg;
	    signal(SIGPIPE, SIG_IGN);
	    break;
 	case 't':
	    if ((response_time = atoi(optarg)) <= 0)
		usage("invalid timeout");
	    break;
	case 'u':
	    want_err = EHOSTUNREACH;
	    break;
	case 'U':
	    want_err = ~EHOSTUNREACH;
	    break;
	case 'v':
	    verbose = 1;
	    break;
	case 'w':
	    if ((banner_time = atoi(optarg)) <= 0)
		usage("invalid timeout");
	    break;
	default:
	    usage((char *) 0);
	    break;
	}
    }
    argc -= (optind - 1);
    argv += (optind - 1);
    if (argc < 3)
	usage("missing host or service argument");

    socket_to_port = NEW(int, open_file_limit);

    FD_ZERO(&write_socket_mask);
    FD_ZERO(&read_socket_mask);
    ports_busy = 0;

    /*
     * Allocate the socket to read ICMP replies.
     */
    if ((pe = getprotobyname("icmp")) == 0)
	error("icmp: unknown protocol");
    if ((icmp_sock = socket(AF_INET, SOCK_RAW, pe->p_proto)) < 0)
	error("icmp socket: %m");
    FD_SET(icmp_sock, &read_socket_mask);

    /*
     * Scan the ports.
     */
    memset((char *) &sin, 0, sizeof(sin));
    sin.sin_addr = find_addr(argv[1]);
    sin.sin_family = AF_INET;

    if (response_time > 0)
	alarm(response_time);

    for (ports = argv + 2; *ports; ports++)
	scan_ports(*ports);
    while (ports_busy > 0)
	monitor_ports(WAIT);

    return (0);
}

/* usage - explain command syntax */

usage(why)
char   *why;
{
    if (why)
	remark(why);
    error("usage: %s [-abuU] [-l load] [-w time] host ports...", progname);
}

/* scan_ports - scan ranges of ports */

scan_ports(service)
char   *service;
{
    char   *cp;
    int     min_port;
    int     max_port;
    int     port;
    int     sock;

    /*
     * Translate service argument to range of port numbers.
     */
    if ((cp = strchr(service, '-')) != 0) {
	*cp++ = 0;
	min_port = (service[0] ? ntohs(find_port(service, "tcp")) : 1);
	max_port = (cp[0] ? ntohs(find_port(cp, "tcp")) : 65535);
    } else {
	min_port = max_port = ntohs(find_port(service, "tcp"));
    }

    /*
     * Iterate over each port in the given range. Try to keep as many sockets
     * open at the same time as possible. Gradually increase the number of
     * probes so that they will be spread in time.
     */
    for (port = min_port; port <= max_port; port++) {
	if (verbose)
	    remark("connecting to port %d", port);
	sin.sin_port = htons(port);
	while ((sock = socket(sin.sin_family, SOCK_STREAM, 0)) < 0) {
	    remark("socket: %m");
	    monitor_ports(WAIT);
	}
	add_socket(sock, port);
	non_blocking(sock, YES);
	if (connect(sock, (struct sockaddr *) & sin, sizeof(sin)) < 0
	    && errno != EINPROGRESS) {
	    report_and_drop_socket(sock, errno);
	    continue;
	}
	if (ports_busy < load_limit && ports_busy < ports_done) {
	    monitor_ports(NOWAIT);
	} else {
	    while (ports_busy >= load_limit || ports_busy >= ports_done)
		monitor_ports(WAIT);
	}
    }
}

/* monitor_ports - watch for socket activity */

monitor_ports(wait)
int     wait;
{
    fd_set  read_mask;
    fd_set  write_mask;
    static struct timeval waitsome = {1, 1,};
    static struct timeval waitnot = {0, 0,};
    int     sock;
    char    ch;

    if (banner_info == 0) {

	/*
	 * When a connect() completes, report the socket and get rid of it.
	 */
	write_mask = write_socket_mask;
	read_mask = read_socket_mask;
	if (select(max_sock + 1, &read_mask, &write_mask, (fd_set *) 0,
		   wait ? (struct timeval *) 0 : &waitnot) < 0)
	    error("select: %m");
	if (FD_ISSET(icmp_sock, &read_mask))
	    receive_icmp(icmp_sock);
	for (sock = 0; ports_busy > 0 && sock <= max_sock; sock++) {
	    if (FD_ISSET(sock, &write_mask)) {
		if (read(sock, &ch, 1) < 0 && errno != EWOULDBLOCK && errno != EAGAIN) {
		    report_and_drop_socket(sock, errno);
		} else {
		    report_and_drop_socket(sock, 0);
		}
	    }
	}
    } else {

	/*
	 * When a connect() completes, try to receive some data within
	 * banner_time seconds. Assume we have received all banner data when
	 * a socket stops sending for BANNER_IDLE seconds.
	 */
	write_mask = write_socket_mask;
	read_mask = read_socket_mask;
	if (select(max_sock + 1, &read_mask, &write_mask, (fd_set *) 0,
		   wait ? &waitsome : &waitnot) < 0)
	    error("select: %m");
	if (FD_ISSET(icmp_sock, &read_mask))
	    receive_icmp(icmp_sock);
	gettimeofday(&now, (struct timezone *) 0);
	for (sock = 0; ports_busy > 0 && sock <= max_sock; sock++) {
	    if (sock == icmp_sock)
		continue;
	    if (FD_ISSET(sock, &write_mask)) {
		FD_CLR(sock, &write_socket_mask);
		FD_SET(sock, &read_socket_mask);
		banner_info[sock].connect_time = now;
		if (send_string)
		    do_send_string(sock);
	    } else if (FD_ISSET(sock, &read_mask)) {
		switch (read_socket(sock)) {
		case -1:
		    if (errno != EWOULDBLOCK && errno != EAGAIN)
			report_and_drop_socket(sock, errno);
		    break;
		case 0:
		    report_and_drop_socket(sock, 0);
		    break;
		}
	    } else if (FD_ISSET(sock, &read_socket_mask)) {
		if (time_since_connect(sock) > banner_time
		    || time_since_read(sock) > BANNER_IDLE)
		    report_and_drop_socket(sock, 0);
	    }
	}
    }
}

/* read_socket - read data from server */

read_socket(sock)
int     sock;
{
    BANNER_INFO *bp = banner_info + sock;
    unsigned char *cp;
    int     len;
    int     count;

    if (bp->buf == 0)
	bp->buf = (unsigned char *) mymalloc(BANNER_LENGTH);
    cp = bp->buf + bp->count;
    len = BANNER_LENGTH - bp->count;

    if (len == 0)
	return (0);

    bp->read_time = now;

    /*
     * Process banners with one-character reads so that we can detect telnet
     * options.
     */

    if ((count = read(sock, cp, 1)) == 1) {
	if (cp[0] == IAC) {
	    if ((count = read(sock, cp + 1, 2)) == 2) {
		if (cp[1] == WILL || cp[1] == WONT) {
		    cp[1] = DONT;
		    bp->flags |= F_TELNET;
		    write(sock, cp, 3);
		} else if (cp[1] == DO || cp[1] == DONT) {
		    cp[1] = WONT;
		    bp->flags |= F_TELNET;
		    write(sock, cp, 3);
		}
	    }
	} else {				/* cp[0] != IAC */
	    bp->count++;
	}
    }
    return (count);
}

/* report_and_drop_socket - report what we know about this service */

report_and_drop_socket(sock, err)
int     sock;
int     err;
{
    alarm(0);

    if (show_all || want_err == err || (want_err < 0 && want_err != ~err)) {
	struct servent *sp;
	int     port = socket_to_port[sock];

	printf("%d:%s:", port, (sp = getservbyport(htons(port), "tcp")) != 0 ?
	       sp->s_name : "UNKNOWN");
	if (banner_info) {
	    BANNER_INFO *bp = banner_info + sock;

	    if (bp->flags & F_TELNET)
		putchar('t');
	    putchar(':');
	    if (bp->count > 0)
		print_data(stdout, bp->buf, bp->count);
	}
	if (err && show_all)
	    printf("%s", strerror(err));
	printf("\n");
	fflush(stdout);
    }
    drop_socket(sock);
}

/* add_socket - say this socket is being connected */

add_socket(sock, port)
int     sock;
int     port;
{
    BANNER_INFO *bp;

    socket_to_port[sock] = port;
    if (banner_info) {
	bp = banner_info + sock;
	bp->count = 0;
	bp->buf = 0;
	bp->flags = 0;
    }
    FD_SET(sock, &write_socket_mask);
    if (sock > max_sock)
	max_sock = sock;
    ports_busy++;
}

/* drop_socket - release socket resources */

drop_socket(sock)
int     sock;
{
    BANNER_INFO *bp;

    if (banner_info && (bp = banner_info + sock)->buf)
	free((char *) bp->buf);
    close(sock);
    FD_CLR(sock, &read_socket_mask);
    FD_CLR(sock, &write_socket_mask);
    ports_busy--;
    ports_done++;
}

/* time_since_read - how long since read() completed? */

time_since_read(sock)
int     sock;
{
    BANNER_INFO *bp = banner_info + sock;

    return (bp->count == 0 ? 0 : time_since(bp->read_time));
}

/* time_since_connect - how long since connect() completed? */

time_since_connect(sock)
int     sock;
{
    BANNER_INFO *bp = banner_info + sock;

    return (time_since(bp->connect_time));
}

/* receive_icmp - receive and decode ICMP message */

receive_icmp(sock)
int     sock;
{
    union {
	char    chars[BUFSIZ];
	struct ip ip;
    }       buf;
    int     data_len;
    int     hdr_len;
    struct ip *ip;
    struct icmp *icmp;
    struct tcphdr *tcp;
    int     port;

    if ((data_len = recv(sock, (char *) &buf, sizeof(buf), 0)) < 0) {
	error("error: recv: %m");
	return;
    }

    /*
     * Extract the IP header.
     */
    ip = &buf.ip;
    if (ip->ip_p != IPPROTO_ICMP) {
	error("error: not ICMP proto (%d)", ip->ip_p);
	return;
    }

    /*
     * Extract the IP payload.
     */
    hdr_len = ip->ip_hl << 2;
    if (data_len - hdr_len < ICMP_MINLEN) {
	remark("short ICMP packet (%d bytes)", data_len);
	return;
    }
    icmp = (struct icmp *) ((char *) ip + hdr_len);
    data_len -= hdr_len;

    if (icmp->icmp_type != ICMP_UNREACH)
	return;

    /*
     * Extract the offending IP packet header.
     */
    if (data_len < offsetof(struct icmp, icmp_ip) + sizeof(icmp->icmp_ip)) {
	remark("short IP header in ICMP");
	return;
    }
    ip = &(icmp->icmp_ip);
    if (ip->ip_p != IPPROTO_TCP)
	return;
    if (ip->ip_dst.s_addr != sin.sin_addr.s_addr)
	return;

    /*
     * Extract the offending TCP packet header.
     */
    hdr_len = ip->ip_hl << 2;
    tcp = (struct tcphdr *) ((char *) ip + hdr_len);
    data_len -= hdr_len;
    if (data_len < offsetof(struct tcphdr, th_dport) + sizeof(tcp->th_dport)) {
	remark("short TCP header in ICMP");
	return;
    }

    /*
     * Process ICMP subcodes.
     */
    switch (icmp->icmp_code) {
    case ICMP_UNREACH_NET:
    case ICMP_UNREACH_PROTOCOL:
	/* error("error: network or protocol unreachable"); */
	/* NOTREACHED */
    case ICMP_UNREACH_PORT:
    case ICMP_UNREACH_HOST:
	port = ntohs(tcp->th_dport);
	for (sock = 0; sock < open_file_limit; sock++)
	    if (socket_to_port[sock] == port) {
		report_and_drop_socket(sock, EHOSTUNREACH);
		return;
	    }
	break;
    }
}

/* do_send_string - send the send string */

do_send_string(sock)
int     sock;
{
    char    buf[BUFSIZ];
    char   *cp = buf;
    char    ch;
    int     c;
    int     i;
    char   *s = send_string;

    while (*s && cp < buf + sizeof(buf) - 1) {	/* don't overflow the buffer */

	if (*s != '\\') {			/* ordinary character */
	    *cp++ = *s++;
	} else if (isdigit(*++s) && *s < '8') {	/* \nnn octal code */
	    sscanf(s, "%3o", &c);
	    *cp++ = c;
	    for (i = 0; i < 3 && isdigit(*s) && *s < '8'; i++)
		s++;
	} else if ((ch = *s++) == 0) {		/* at string terminator */
	    break;
	} else if (ch == 'b') {			/* \b becomes backspace */
	    *cp++ = '\b';
	} else if (ch == 'f') {			/* \f becomes formfeed */
	    *cp++ = '\f';
	} else if (ch == 'n') {			/* \n becomes newline */
	    *cp++ = '\n';
	} else if (ch == 'r') {			/* \r becomes carriage ret */
	    *cp++ = '\r';
	} else if (ch == 's') {			/* \s becomes blank */
	    *cp++ = ' ';
	} else if (ch == 't') {			/* \t becomes tab */
	    *cp++ = '\t';
	} else {				/* \any becomes any */
	    *cp++ = ch;
	}
    }
    write(sock, buf, cp - buf);
}
